zetaclient/chains/bitcoin/client/client.go

```diff
@@ -0,0 +1,270 @@
+//
+// # ISC License
+//
+// Copyright (c) 2013-2024 The btcsuite developers
+// Copyright (c) 2015-2016 The Decred developers
+//
+// Permission to use, copy, modify, and distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	types "github.com/btcsuite/btcd/btcjson"
+	chains "github.com/btcsuite/btcd/chaincfg"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/tendermint/btcd/chaincfg"
+
+	"github.com/zeta-chain/node/zetaclient/config"
+	"github.com/zeta-chain/node/zetaclient/logs"
+	"github.com/zeta-chain/node/zetaclient/metrics"
+)
+
+// Client Bitcoin RPC client
+type Client struct {
+	hostURL    string
+	client     *http.Client
+	clientName string
+	config     config.BTCConfig
+	params     chains.Params
+	logger     zerolog.Logger
+}
+
+type Opt func(c *Client)
+
+type rawResponse struct {
+	Result json.RawMessage `json:"result"`
+	Error  *types.RPCError `json:"error"`
+}
+
+const (
+	// v1 means "no batch mode"
+	rpcVersion = types.RpcVersion1
+
+	// rpc command id. as we don't send batch requests, it's always 1
+	commandID = uint64(1)
+)
+
+var _ client = (*Client)(nil)
+
+func WithHTTP(httpClient *http.Client) Opt {
+	return func(c *Client) { c.client = httpClient }
+}
+
+// New Client constructor
+func New(cfg config.BTCConfig, chainID int64, logger zerolog.Logger, opts ...Opt) (*Client, error) {
+	params, err := resolveParams(cfg.RPCParams)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to resolve chain params")
+	}
+
+	clientName := fmt.Sprintf("btc:%d", chainID)
+
+	c := &Client{
+		hostURL:    normalizeHostURL(cfg.RPCHost, true),
+		client:     defaultHTTPClient(),
+		config:     cfg,
+		params:     params,
+		clientName: clientName,
+		logger: logger.With().
+			Str(logs.FieldModule, "btc_client").
+			Int64(logs.FieldChain, chainID).
+			Logger(),
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c, nil
+}
+
+// send sends RPC command to the server via http post request
+func (c *Client) sendCommand(ctx context.Context, cmd any) (json.RawMessage, error) {
+	method, reqBody, err := c.marshalCmd(cmd)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to marshal cmd")
+	}
+
+	// ps: we can add retry logic if needed
```

Comment:
imo the best place to do retries in our case is from the top down (orchestrator/scheduler level)